packages/language/src/language-server/types.ts

```diff
@@ -109,14 +113,90 @@ export function severityToLsp(severity: Severity): lsp.DiagnosticSeverity {
 
 export interface Diagnostic {
   severity: Severity;
-  uri: string;
-  range: Range;
+  uri?: string;
+  range?: Range;
   message: string;
   code?: string;
   data?: any;
   source?: string;
 }
 
+export function diagnosticFromCode(
+  code: SimplePLICode,
+  token: Token | null | undefined,
+): Diagnostic;
+export function diagnosticFromCode(
+  code: ParametricPLICode,
+  token: Token | null | undefined,
+  ...args: (string | number | undefined)[]
+): Diagnostic;
+export function diagnosticFromCode(
+  code: SimplePLICode | ParametricPLICode,
+  token: Token | null | undefined,
+  ...args: (string | number | undefined)[]
+): Diagnostic {
+  let uri: string | undefined = undefined;
+  let range: Range | undefined = undefined;
+  if (token) {
+    uri = tokenToUri(token);
+    range = tokenToRange(token);
+  }
+  let message = code.message;
+  if (typeof message === "string") {
+    // Simple message, no parameters
+    return {
+      severity: code.severity,
+      uri,
+      range,
+      message,
+      code: code.fullCode,
+    };
+  } else {
+    // Parametric message, format with args
+    return {
+      severity: code.severity,
+      uri,
+      range,
+      message: message(...args),
+      code: code.fullCode,
+    };
+  }
+}
+
+export function diagnostic(
+  severity: Severity,
+  message: string,
+  token: Token | null | undefined,
+): Diagnostic;
+export function diagnostic(
+  severity: Severity,
+  message: string,
+  range: Range | null | undefined,
```

Comment:
Why NULL and UNDEFINED? Also for URI and TOKEN?

Reply:
Because in some cases, the ranges/URIs are simply unavailable (mainly due to the preprocessor or invalid AST). Instead of checking for them **everywhere**, we simply accept those as valid values and filter them out at the end before sending them to the user.